### PR TITLE
MCMC kernel for SCITE inference

### PR DESCRIPTION
**Pull request by gordonkoehn**
Implementing the simplest MCMC kernel from the SCITE paper (Jahn et al. 2016), assuming that the false negative and false positive rates are known. - based on PR #10 by @pawel-czyz.

### CODE REVIEW

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 1]
```diff
@@ -23,22 +26,27 @@ class Tree:
...
     tree_topology: jax.Array
     labels: jax.Array
 
+    def to_TreeNode(self) -> TreeNode:
+        """Converts this Tree to a TreeNode.
+        Returns the root node of the tree."""
+        root = tree_inf.adjacency_to_root_dfs(self.tree_topology)
+        return root
+
+    # TODO: Pawel / Gordon: discuss Tree vs TreeNode - merge / interface / etc.
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_mcmc.py:48]: The idea here is that `TreeNode` is in the public API and it could be used with multiple subpackages.

In this subpackage we have `Tree`, which is *an internal (private) representation* of a `TreeNode` in the format which makes it easy to do MCMC: namely, adjacency matrix encoding topology and the labels, determining the mapping from the tree topology to the mutations. It should be a part of the private API (i.e., we don't export it to any "visible" `__init__.py`) and, yes, conversion between these should be possible (but only for our internal purposes).

Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_mcmc.py:48]: OK, understood - makes sense. I will revert the `Tree` class back to private. A conversion is already implemented from `Tree` -> `TreeNode` mainly for easy visualization to check workings / debug.

Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_mcmc.py]: Wonderful! :slightly_smiling_face:

File: src/pyggdrasil/tree_inference/_tree.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    - adj_matrix: a JAX array of shape (n, n) representing the adjacency matrix
+    - labels: a JAX array of shape (n,) representing the labels of the nodes
+    - parent: an integer representing
+        the label of the node whose descendants we want to find
+
+    Returns:
+    - a JAX array of integers representing the labels of the descendants of node parent
+      in order of nodes in the adjacency matrix, i.e. the order of the labels
+      if includeParent is True, the parent is included in the list of descendants
+    """
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_tree.py:127]: `_mcmc.py` currently used  this`get_descendants()` - a second implementation using Floyd-Warshall exists in `_simulate.py`

File: src/pyggdrasil/tree_inference/_tree.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            return ~(jnp.array_equal(prev_matrix, curr_matrix)) and (exp_counter <= exp)
+
+        def loop_body_fn(carry):
+            prev_matrix, curr_matrix = carry
+            new_matrix = jnp.dot(curr_matrix, bool_mat)
+            return curr_matrix, new_matrix
+
+        (_, adj_matrix_exp), _ = jax.lax.while_loop(
+            loop_cond_fn, loop_body_fn, (bool_mat, bool_mat)
+        )
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_tree.py:186]: A failed attempt to stop exponentiation if no changes occurred, in the last step. - may remove.

**Review by gordonkoehn — COMMENTED** [review]
Implemented all tree moves, proposals and understood/corrected your implementation of the mcmc kernel. 

This PR took me a little longer. I got lost in implementing `get_descendants` using matrix exponentiation and did go down the wrong route for the `prune_and_reattach_move` at first.

Thus there are many files here. All important for the MCMC is in `_mcmc.py`, `_tree.py`.  - `_mcmc_util.py` contains my blunt first attempt at `prune_and_reattach_move`.

File: tests/tree_inference/test_mcmc.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+    assert jnp.array_equal(new_tree.tree_topology, new_tree_corr.tree_topology)
+    assert jnp.array_equal(new_tree.labels, new_tree_corr.labels)
+
+
+# TODO: test _swap_node_labels_proposal
+
+# TODO: test _prune_and_reattach_proposal
+
+# TODO: test _swap_subtrees_proposal
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply tests/tree_inference/test_mcmc.py:200]: Perhaps testing it should go into another PR (I'm not sure how important it is, but perhaps worth opening an issue).

Several testing strategies:
1. For `_swap_node_labels_proposal`: if we have root + 3 nodes and sample 1,000 times, will we arrive at (approximately) 1000/3! samples in each category? Is the root always preserved to be the last one?
2. For `_prune_and_reattach` and `swap_subtrees_proposal`: do we always end up with a cycle-less tree? Is the root never pruned/swapped?

**Review by pawel-czyz — APPROVED** [review]
Excellent work! I think a lot of these will be useful for a JAX sampler compatible with BlackJAX.

As a general advice (it wasn't in the scope of this PR, you've exceeded the expectations), when you program in JAX, you may want to try (a) to avoid `if` statements and exceptions (using `jax.lax.cond` if needed. Exceptions are good in programming, but aren't really compatible with JAX's JIT (sadly...)), (b) to use arrays of fixed size (as for different sizes JAX needs to compile them again, what takes time).

I'd suggest to take another look at the $\pm \Delta_q$ comment (I responded to it, let me know what you think) and when we're sure that the sign is right, to merge it.

With this, I think we're good to go with implementing the likelihood function and then we can have the sampling loop! :slightly_smiling_face:

File: src/pyggdrasil/tree_inference/_tree.py (hunk 3 above)
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_tree.py:186]: Removed.

File: src/pyggdrasil/tree_inference/_tree.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    """Returns the root label of a tree
+
+    Args:
+        tree: Tree
+            tree to get root label of
+
+    Returns:
+        root_label: int
+            root label of the tree
+    """
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_tree.py:227]: Oh - my apologies, it works just fine. I meant it was merely used to debug other code.

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 6]
```diff
@@ -65,12 +54,29 @@ def _prune_and_reattach_proposal(
...
     """
-    raise NotImplementedError
+    # get random keys
+    rng_prune, rng_reattach = random.split(key)
+    # pick a random non-root node to prune
+    pruned_node = int(random.choice(rng_prune, tree.labels[:-1]))
+    # get descendants of pruned node
+    descendants = tr._get_descendants(tree.tree_topology, tree.labels, pruned_node)
+    # possible nodes to reattach to - including pruned node for aperiodic case
+    possible_nodes = jnp.setdiff1d(tree.labels, descendants)
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_mcmc.py:71]: Oh, I see. Added a TODO to modify later pahaps.

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 7]
```diff
@@ -211,12 +282,13 @@ def _mcmc_kernel(
...
+    # log p(proposal)
     logprob_proposal = logprobability_fn(proposal)
 
     # This is the logarithm of the famous Metropolis-Hastings ratio:
-    # https://en.wikipedia.org/wiki/Metropolis%E2%80%93Hastings_algorithm#Formal_derivation
-    # TODO(Pawel, Gordon): Triple check this.
-    log_ratio = logprob_proposal - logprobability - log_q_diff
+    # log A (proposal | tree) = log p(proposal) - log p(tree)
+    #                           + (log q(proposal | tree) - log q(proposal | tree))
+    log_ratio = logprob_proposal - logprobability + log_q_diff
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_mcmc.py:291]: Yes, you are right - reverted the changes. And found/fixed a related issue. 

My confusion originated from a mistake I made implementing '_swap_subtrees_proposal'. The function used to return
$$\Delta q = \mathbf{-} \log q(new | old) \mathbf{+} \log q(old | new)$$
despite its docstring saying the correct, as you wrote.

I made this mistake as that's the way it is written in the SCITE paper Supplementary p15 top. 
$$\frac{q(T|T')}{q(T'|T)} = \frac{d(k)+1}{d(i)+1}$$

with $k$ the descendant of $i$. What we want is the inverse log of this.

File: src/pyggdrasil/tree_inference/_tree.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+          Note: the last position always maps to itself,
+          as it's the root, and we use the convention
+          that root has the largest index.
+          Shape ``(N+1,)``
+    """
+
+    tree_topology: jax.Array
+    labels: jax.Array
+
+    def to_TreeNode(self) -> TreeNode:
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_tree.py:41]: Ok, noted. Will leave it for now.

File: src/pyggdrasil/tree_inference/_mcmc_util.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
+"""Utility Functions for _mcmc.py
+"""
+import jax.numpy as jnp
+import jax.scipy as jsp
+
+from pyggdrasil.tree_inference._tree import Tree
+import pyggdrasil.tree_inference._tree as tr
+
+
+# TODO: consider moving all these 3 function to tests only
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_mcmc_util.py:10]: Ok, then I'll leave the functions where they are now. Thanks.

File: src/pyggdrasil/tree_inference/_mcmc.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_mcmc.py]: Thanks for the update! Great you resolved it :slightly_smiling_face:
